src/main/scala/uk/co/gresearch/spark/diff/DiffComparator.scala

```diff
@@ -0,0 +1,118 @@
+    override def description: String = "Compares two numbers and considers them equal as long as their difference is below the given divergence."
+    override def compare(left: Column, right: Column): Expression = {
+      Or(And(IsNull(left.expr), IsNull(right.expr)),
+      LessThanOrEqual(abs(left - right).expr, lit(divergence + significantDigitAddon).expr))
+    }
+  }
+
+  /**
+   * Fuzzy timestamp comparator allowing for a grace duration before and after the right column values
+   * @param leftBeforeRight - number of seconds before the right timestamp in which the left timestamp may occur to be considered equal
+   * @param leftAfterRight - number of seconds after the right timestamp in which the left timestamp may occur to be considered equal
+   * @param leftTimeZone - the time zone of the left column (by default: UTC)
+   * @param rightTimeZone - the time zone of the left column (by default: UTC)
+   */
+  case class FuzzyDateComparator(
+    leftBeforeRight: Int,
+    leftAfterRight: Int,
+    leftTimeZone: String = "UTC",
+    rightTimeZone: String = "UTC"
+  ) extends DiffComparator(Seq[String]()){
+    assert(leftBeforeRight >= 0 && leftAfterRight >= 0, "Durations below one second are not allowed for FuzzyDateComparator.")
+    override def description: String = "For comparing two timestamps allowing for a given diversion before and after "
+
+    private val defaultUnixFormat = Literal("yyyy-MM-dd HH:mm:ss") // is actually not used
+    private val leftTz = Option(getTimeZone(leftTimeZone)).getOrElse(throw new IllegalArgumentException("Invalid time zone: " + leftTimeZone))
+    private val rightTz = Option(getTimeZone(rightTimeZone)).getOrElse(throw new IllegalArgumentException("Invalid time zone: " + rightTimeZone))
+
+    /**
+     * The Compare Expression of two other expressions (usually resolved as instance of [[BinaryExpression]]
+     *
+     * @param left  - the left expression of a binary expression (left Column)
+     * @param right - the right expression of a binary expression (right Column)
+     */
+    override def compare(left: Column, right: Column): Expression = {
+      val leftTimestamp = to_utc_timestamp(left, leftTz.getID)
+      val rightTimestamp = to_utc_timestamp(right, rightTz.getID)
+      val diffInSeconds = new Column(Subtract(
+        UnixTimestamp(leftTimestamp.expr, defaultUnixFormat, Some(leftTz.getID)),
+        UnixTimestamp(rightTimestamp.expr, defaultUnixFormat, Some(rightTz.getID))
+      ))
+
+      ( leftTimestamp.isNull && rightTimestamp.isNull ||
+        diffInSeconds > lit(0) && diffInSeconds <= lit(leftBeforeRight) ||
+        diffInSeconds <= lit(0) && abs(diffInSeconds) <= lit(leftAfterRight)).expr
+    }
+}
+
+  // this is the default comparator
+  val EqualNullSafeComparator: DiffComparator = new DiffComparator(Seq("<=>", "eqNullSafe")){
```

Comment:
Not sure about how useful these comparators are, how common or how much edge-case they would be. The null safe equality is central to correct diffing, otherwise null values would always flag up as changes.